contributors/design-proposals/storage/volume-topology-scheduling.md

```diff
@@ -266,41 +304,502 @@ const (
+* There is a gap during cluster initialization where the StorageClass object
+  has been created, and the provisioner has not reported capacity yet. In this
+  scenario, the provisioner should fail provisioning until it has reported its
+  capacity and not process any PVCs that have not been updated with the chosen
+  topology value.
+
+#### Local Volume Example
+This example specifies that node1 has 100Gi, and node2 has 200Gi available
+for dynamic provisioning. All other nodes have 0Gi available for dynamic
+provisioning.
+```
+apiVersion: storage.k8s.io/v1
+kind: StorageClass
+metadata:
+  name: standard
+...
+status:
+  capacity:
+    topologyKey: kubernetes.io/hostname
+    capacities:
+      node1: 100Gi
+      node2: 200Gi
+```
+
+TODO: should this be in the local volume design doc instead?
+Some downsides specific to the local volume implementation:
+* StorageClass object size can be big for large clusters. Assuming that
+  each node entry in the Capacities map could take up around 100 bytes
+  (size of node name + resource.Quantity), a 1000 node cluster that uses
+  dynamic provisioning of local volumes could add 100Ki to the StorageClass
+  object.
+* If each node has a local volume provisioner, there can be many
+  StorageClass update conflicts if many try to update their capacity at the
+  same time, especially during cluster initialization. This could be mitigated
+  by having each provisioner sleep for a random amount of time during
+  startup.
+* A security concern is that each local volume provsioner could potentially
+  modify the capacity reported by another node. This could cause
+  provisioning to fail if a node does not have enough capacity, or it could
+  cause the scheduler to skip nodes due to no capacity. This could be
+  mitigated when per-node identity is supported and a validating webhook
+  could be provided.
+
+
+#### Multi-dimensional Example
+This example specifies that rack1 in zoneA has 100Gi, and rack1 in zoneB has
+200Gi available for dynamic provisioning. All other racks have 0Gi available
+for dynamic provisioning.
+```
+apiVersion: storage.k8s.io/v1
+kind: StorageClass
+metadata:
+  name: standard
+...
+status:
+  capacity:
+    topologyKey: zone,rack
+    capacities:
+      zoneA,rack1: 100Gi
+      zoneB,rack1: 200Gi
+```
+
+#### Relationship to StorageClassQuota
+Although both the new StorageClassCapacity field and existing StorageClassQuota
+manage storage consumption, they address different use cases.
+
+The StorageClassCapacity field is actual available capacity per topology domain
+reported by the storage system, while the StorageClassQuota represents policies
+set by the cluster admin and does not take into account topology domain.
+
+For example, a storage system is configured to support 200Gi in zone1 for a
+StorageClass that is shared by multiple users.  An administrator can use
+StorageClassQuota to restrict capacity usage for each user.
+
+## API Upgrades
+
+### Zonal PV Upgrade
```

Comment:
Add an example for multi-zonal PV upgrade because the zone label value format is special.